39 - Module restructuring and rearchitecture

GPortas opened the pull request:
## What this PR does / why we need it:

This PR completely restructures the module, based on a new architecture design.

The code now only includes a single Dataverse API operation, which serves as an end-to-end architecture test, thus losing the support for the operations offered in the old package. These operations, among others, will be implemented in the upcoming PRs before releasing the package, to make the next release version (`2.0.0`) backwards compatible with the `1.x.x` versions.

## Which issue(s) this PR closes:

- Closes https://github.com/IQSS/dataverse-frontend/issues/39

## Special notes for your reviewer:

### Testing

Includes both unit tests and an integration test related to the added functionality. The integration test is very basic and currently points to `demo.dataverse.org`. I have added a _TODO_ mentioning that it is necessary to establish a suitable environment for integration tests (Idea: we could setup that environment with containers).

Different test suites have been configured to be executed all together, or independently, depending on the test nature (unit or integration), so there should be no problem in including only unit tests execution in future code validation mechanisms, such as hooks or PR checks. You can find more information about this in the README.

### Docs

The old docs, based on TypeDoc, have been removed along with the old code. In order not to extend this PR, a specific issue will be created to address the documentation of the module, since by having to recreate the documentation from scratch and given the age of the old documentation, it may be a good opportunity to explore other options for documenting.

## Suggestions on how to test this:

### Tests and linting

There are instructions in the README on how you can run the different types of tests, coverage, as well as other linting and formatting related operations.

### Testing the functionality

As mentioned, the restructured package only covers one functionality at the moment, which is retrieving the Dataverse version.

To test this functionality we have to import the npm package from a test application. Usually we would install the package from the npm registry, but since it is a development unpublished version, we have to reference a pre-built local package from the test application.

To build the package locally:

- 1) Install dependencies omitting dev: `npm install --omit=dev`.
- 2) Build the package code: `npm run build`.
- 3) Create the .tgz local package: `npm pack`.

In [dataverse-frontend/39-js-dataverse-test](https://github.com/IQSS/dataverse-frontend/tree/39-js-dataverse-test) there is an example of how the local package is imported and configured and a call to "getDataverseVersion" is made. In this example we can see:

- https://github.com/IQSS/dataverse-frontend/blob/39-js-dataverse-test/package.json#L105  - The package.json dependency setup
- https://github.com/IQSS/dataverse-frontend/blob/39-js-dataverse-test/src/sections/hello-dataverse/HelloDataverse.tsx#L11 - Call to `getDataverseVersion` from the UI
- https://github.com/IQSS/dataverse-frontend/blob/39-js-dataverse-test/index.html#L35 - API base URL environment variable setup

## Is there a release notes update needed for this change?:

N/A

## Additional documentation:

- js-dataverse old code base analysis: [link](https://docs.google.com/document/d/1s_ZPr3BRlCuL5Jch8qJ3odncIgkmo6krSn2tMVv5e44)

GPortas committed:
Changed: updated dependencies and old package files moved to temporal 'old' folders

GPortas committed:
Changed: package.json description

GPortas committed:
Removed: unused tsfmt.json file

GPortas committed:
Added: domain use case logic with initial sample use case

GPortas committed:
Added: use case error handling with unit tests and WIP of the repository implementation

GPortas committed:
Added: DataverseInfoRepository tests

GPortas committed:
Added: missing '^' prefix to package.json versions

GPortas committed:
Added: parametrized API URL for repository impl

GPortas committed:
Added: specific test type scripts (test:unit and test:integration)

GPortas committed:
Refactor: test response extracted to variable

GPortas committed:
Added: index.ts for 'info' module exports

GPortas committed:
Added: package modules initialization

GPortas committed:
Removed: old package code

GPortas committed:
Removed: unnecessary packages

GPortas committed:
Changed: updated README

GPortas committed:
Removed: old test path ignore from jest configs

GPortas committed:
Changed: issue and PR templates

GPortas committed:
Added: src index.ts

GPortas committed:
Removed: old docs

GPortas committed:
Fixed: missing package.json dependencies added

GPortas committed:
Refactor: removed result model and handling errors with exceptions on module use case calls

GPortas committed:
Added: DataverseVersion model to include both version and build numbers

pdurbin commented on README.md:
```diff
@@ -1,71 +1,12 @@
-`public async getDatasetFiles(datasetId: string): Promise<AxiosResponse> {`
+Module available as `js-dataverse` at https://www.npmjs.com/package/js-dataverse
 
 ## Build project
```
Can we please specify which version of node to use? Or if the node version is in a config file, it's fine to reference that file instead.

pdurbin commented on README.md:
```diff
@@ -75,33 +16,47 @@ In order to build the project, we need to run the following command:
 
```
I'm getting these errors:

```
$ yarn build
yarn run v1.22.19
$ tsc
/bin/sh: tsc: command not found
error Command failed with exit code 127.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
$ 
$ npm run build

> js-dataverse@2.0.0 build
> tsc

sh: tsc: command not found
$ 
```

If there's an extra command to run, we should document it.

I tried running `npm install typescript -g` which is a popular answer at https://stackoverflow.com/questions/39404922/tsc-command-not-found-in-compiling-typescript/39413342#39413342

Now I'm getting this error:

```
$ yarn build
yarn run v1.22.19
$ tsc
src/info/index.ts:4:81 - error TS2580: Cannot find name 'process'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.

4 const getDataverseVersion = new GetDataverseVersion(new DataverseInfoRepository(process.env.DATAVERSE_API_URL));
                                                                                  ~~~~~~~

src/info/infra/repositories/DataverseInfoRepository.ts:2:38 - error TS2307: Cannot find module 'axios' or its corresponding type declarations.

2 import axios, { AxiosResponse } from 'axios';
                                       ~~~~~~~

src/info/infra/repositories/DataverseInfoRepository.ts:14:76 - error TS7006: Parameter 'error' implicitly has an 'any' type.

14     const response = await axios.get(`${this.apiUrl}/info/version`).catch((error) => {
                                                                              ~~~~~


Found 3 errors in 2 files.

Errors  Files
     1  src/info/index.ts:4
     2  src/info/infra/repositories/DataverseInfoRepository.ts:2
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
$ 
```

Similar errors from `npm run build`:

```
$ npm run build

> js-dataverse@2.0.0 build
> tsc

src/info/index.ts:4:81 - error TS2580: Cannot find name 'process'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.

4 const getDataverseVersion = new GetDataverseVersion(new DataverseInfoRepository(process.env.DATAVERSE_API_URL));
                                                                                  ~~~~~~~

src/info/infra/repositories/DataverseInfoRepository.ts:2:38 - error TS2307: Cannot find module 'axios' or its corresponding type declarations.

2 import axios, { AxiosResponse } from 'axios';
                                       ~~~~~~~

src/info/infra/repositories/DataverseInfoRepository.ts:14:76 - error TS7006: Parameter 'error' implicitly has an 'any' type.

14     const response = await axios.get(`${this.apiUrl}/info/version`).catch((error) => {
                                                                              ~~~~~


Found 3 errors in 2 files.

Errors  Files
     1  src/info/index.ts:4
     2  src/info/infra/repositories/DataverseInfoRepository.ts:2
$ 
```

As prompted, I ran `npm i --save-dev @types/node`.

Now the build is working:

```
$ yarn build
yarn run v1.22.19
$ tsc
✨  Done in 3.85s.
$ ls dist
core		index.d.ts	index.js	index.js.map	info
```

Let's document these two commands.

GPortas replied:
I think you shouldn't have had these problems if you execute `npm install` from the start before running any other command (The dependencies you are installing are already included in the `package.json `dependencies).

I have updated the README because it was not specified correctly that it is necessary to execute `npm install`  before `npm run build.`

pdurbin replied:
Yes! Worked fine! Thanks!

pdurbin commented on README.md:
```diff
@@ -75,33 +16,47 @@ In order to build the project, we need to run the following command:
+
 Make sure that you install all the project dependencies
 
 `yarn install` or `npm install`
```
First of all, it's a bit confusing to me to see multiple ways of running the same command. I don't have a strong preference between yarn and npm. Can we pick one? I realize that this a pre-existing condition from the old README! 😄 

When I run `yarn install` it even tells me I should pick one:

`It is advised not to mix package managers in order to avoid resolution inconsistencies caused by unsynchronized lock files.`

Full output:

```
$ yarn install
yarn install v1.22.19
warning package-lock.json found. Your project contains lock files generated by tools other than Yarn. It is advised not to mix package managers in order to avoid resolution inconsistencies caused by unsynchronized lock files. To clear this warning, remove package-lock.json.
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 🔨  Building fresh packages...
success Saved lockfile.
✨  Done in 7.24s.
$ 
```

As I go through the README I guess I'll use yarn since it was listed first but in the frontend repo we are using npm so maybe npm should be used instead for consistency.

GPortas replied:
I agree. I kept both commands from the old README, but it's better to be consistent with npm, as you say.

pdurbin commented on README.md:
```diff
@@ -75,33 +16,47 @@ In order to build the project, we need to run the following command:
 
+Running a linting check on the code:
+
+`yarn lint` or `npm run lint`
```
`yarn lint` found one problem:

```
$ yarn lint
yarn run v1.22.19
$ eslint src --ext .js,.ts

/Users/pdurbin/github/iqss/dataverse-client-javascript/src/core/domain/useCases/UseCase.ts
  2:20  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

✖ 1 problem (0 errors, 1 warning)

✨  Done in 1.79s.
$ 
```

(`yarn lint:fix` didn't fix it by the way.)

GPortas replied:
It is a well-detected warning, following what the typescript-eslint plugin mentions in [https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-explicit-any.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-explicit-any.md).

In my opinion, the use of "any" is detecting, which is the only one in the code, is justified since we have to provide a mechanism for the Use Case base interface to accept as many parameters of any type as the implementations need. See the line: [https://github.com/IQSS/dataverse-client-javascript/blob/ba18a0a58a5648328382c35bbd2d863ade94d415/src/core/domain/useCases/UseCase.ts#L2](https://github.com/IQSS/dataverse-client-javascript/blob/ba18a0a58a5648328382c35bbd2d863ade94d415/src/core/domain/useCases/UseCase.ts#L2).

I couldn't find any other way to implement this, other than using "any". I'll be happy to refactor it if we find a better approach.

Thanks for noticing.

pdurbin commented on .github/ISSUE_TEMPLATE/feature_request.md:
```diff
@@ -0,0 +1,9 @@
+## Overview of the Feature Request
```
```suggestion
## Overview of the feature request
```

Also, should this be `#` (h1)?

GPortas replied:
Changed to lowercase. See my other comment about `##`.

pdurbin replied:
Looks good.

pdurbin commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -0,0 +1,19 @@
+# What steps does it take to reproduce the issue?
```
Should this be the h1? The feature request template has "overview" as the first line.

GPortas replied:
I have decided to use `##` (h2) in all section titles of the templates. So the titles are not too big. This is how I've often seen it in Dataverse issues.

pdurbin commented on package.json:
```diff
@@ -1,7 +1,7 @@
   "name": "js-dataverse",
-  "version": "1.0.48",
-  "description": "A Dataverse module for JavaScript/TypeScript",
+  "version": "2.0.0",
```
I don't know what the best practice is but should we consider using something like 2.0.0-rc.1 so we can actually push the package to npm without it being the official 2.0.0 release?

GPortas replied:
I'm not sure if I would push non-stable builds to npm. It seems safer to me to use local versions for the moment to test the package (it may even be faster). We can talk more about this in the frontend sync meeting.

pdurbin replied:
I guess the version doesn't matter so much until we actual publish to npm. We can talk about it later and change the version later, if need be.

pdurbin commented on test/integration/info/DataverseInfoRepository.test.ts:
```diff
@@ -0,0 +1,12 @@
+
+describe('getDataverseVersion', () => {
+  // TODO: Change API URL to another of an integration test oriented Dataverse instance
+  const sut: DataverseInfoRepository = new DataverseInfoRepository('https://demo.dataverse.org/api/v1');
```
I think for now having the demo server here is fine but we should talk more about the longer term plan for testing.

This comes up for all client libraries! 😄

pdurbin replied:
The conversation about API testing continues in Containerization Working Group meetings: https://ct.gdcc.io

pdurbin commented on package.json:
```diff
@@ -32,25 +32,22 @@
   },
   "homepage": "https://github.com/IQSS/dataverse-client-javascript#readme",
```
When should we rename to repo to https://github.com/IQSS/js-dataverse ?

GPortas replied:
Let's change it when we release the next version to npm. That way we don't affect the links to the repository that currently appear at [https://www.npmjs.com/package/js-dataverse](https://www.npmjs.com/package/js-dataverse).

What do you think?

pdurbin replied:
That's fine but it would also be fine to rename it now. GitHub does a good job of redirecting. I'm resolving this but we can rename it any time, I'd say.

pdurbin reviewed (commented):
Overall, this looks great! I'm excited about the plan to simply create a 2.0.0 version from this repo. I left some comments.

GPortas committed:
Changed: README for using only npm and showing node version

GPortas committed:
Changed: issue templates format

GPortas committed:
Changed: README to better specify deps installation

mreekie commented:
sprint:
- Phil looks at comments
- Bob also needs some time to look at it.
- sized as: 3

MellyGray commented on src/core/domain/repositories/ReadError.ts:
```diff
@@ -0,0 +1,9 @@
+  constructor(reason?: string) {
+    let message = 'There was an error when reading the resource.';
+    if (reason) {
+      message += ' Reason was: ' + reason;
```
```suggestion
      message += ` Reason was: ${reason}`
```

MellyGray commented on src/info/infra/repositories/DataverseInfoRepository.ts:
```diff
@@ -0,0 +1,27 @@
+
+  constructor(apiUrl: string) {
+    this.apiUrl = apiUrl;
+  }
```
```suggestion
  constructor(private readonly apiUrl: string) {}
```

in TS you can declare and initialise at the same time the class attributes

GPortas replied:
That's great, much simpler. Thanks!

MellyGray commented on src/info/infra/repositories/DataverseInfoRepository.ts:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  public async getDataverseVersion(): Promise<DataverseVersion> {
+    const response = await axios.get(`${this.apiUrl}/info/version`).catch((error) => {
+      throw new ReadError(error.response.status + error.response.data ? ': ' + error.response.data.message : '');
+    });
```
```
const response = await axios.get(`${this.apiUrl}/info/version`)
                      .then((response) => getVersionFromResponse(response))
                      .catch((error) => throw new ReadError(error.message));

```

I really prefer being able to read the whole path sequentially. then.then.then.catch

GPortas replied:
I agree, more visual!

MellyGray commented on src/info/infra/repositories/DataverseInfoRepository.ts:
```diff
@@ -0,0 +1,27 @@
+
+  public async getDataverseVersion(): Promise<DataverseVersion> {
+    const response = await axios.get(`${this.apiUrl}/info/version`).catch((error) => {
+      throw new ReadError(error.response.status + error.response.data ? ': ' + error.response.data.message : '');
```
I think we'll need something like` await new Api.get(`${this.apiUrl}/info/version`) `because we don't want to be doing this check `error.response.status + error.response.data ? ': ' + error.response.data.message : '') `everytime we do an API call 

something like:


```
const response = await new Api.get(`${this.apiUrl}/info/version`).catch((error) => {
       throw new ReadError(error.message);
})

```

GPortas replied:
Yes, something like a data source/API client object for the repository, encapsulating API requests and error handling. In any case, I already had these improvements in mind to avoid duplication and improve error handling (which is now very basic), as we add more API operations.

For now, I think it's fine. We will do emergent design on aspects like this for sure!

GPortas committed:
Refactor: using private readonly in constructor

GPortas committed:
Refactor: axios path sequental reading

tainguyenbui commented:
loving these conversations ❤️, great time for me to learn too, thank you

MellyGray commented on src/info/infra/repositories/DataverseInfoRepository.ts:
```diff
@@ -4,17 +4,19 @@ import { ReadError } from '../../../core/domain/repositories/ReadError';
-      throw new ReadError(error.response.status + error.response.data ? ': ' + error.response.data.message : '');
-    });
-    return this.getVersionFromResponse(response);
+    let datasetVersion;
+    await axios
+      .get(`${this.apiUrl}/info/version`)
+      .then((response) => {
+        datasetVersion = this.getVersionFromResponse(response);
+      })
+      .catch((error) => {
+        throw new ReadError(error.response.status + error.response.data ? ': ' + error.response.data.message : '');
+      });
+    return datasetVersion;
```
```suggestion
    return await axios
      .get(`${this.apiUrl}/info/version`)
      .then((response) => this.getVersionFromResponse(response))
      .catch((error) => throw new ReadError(error.response.status + error.response.data ? ': ' + error.response.data.message : ''));
```

If you are not going to use the datasetVersion for other operations you can directly do the return right?

GPortas replied:
The intention was to make it more visual. But I agree, I have simplified it as you propose.

MellyGray replied:
Your intention makes all the sense to me. But I think JS developers are more used to this kind of `return function.then.then.catch` syntax than to putting the return at the end

I don't know if that makes for you, is more javascript, like when people say is more pythonic but with JS, is more javascripthonic (?

Thanks for accepting the change!

GPortas committed:
Refactor: removed unnecessary variable

GPortas committed:
Refactor: string concat in ReadError

GPortas committed:
Changed: package-lock

pdurbin reviewed (dismissed):
Looking good now. Thanks, @GPortas! And thanks also to @MellyGray for suggesting ways to make the library better!

pdurbin committed:
improve README.md and CONTRIBUTING.md #39

- remove yarn
- consistency fixes
- link to upstream Code of Conduct
- explain how to get in touch, how to get help
- fix links between README and CONTRIBUTING to work locally

pdurbin commented:
@GPortas I just pushed the tweak to the docs I mentioned: 8661503

Please take a look and revert or fix up as necessary! Thanks!

kcondon commented:
Having trouble building locally:
npm run build

> js-dataverse@1.0.48 build
> tsc

sh: tsc: command not found


What am I doing wrong? Probably something basic.

tainguyenbui commented:
> Having trouble building locally: npm run build
> 
> > js-dataverse@1.0.48 build
> > tsc
> 
> sh: tsc: command not found
> 
> What am I doing wrong? Probably something basic.

You may not have the typescript installed in your machine.

Here you can find info on how to do it https://www.npmjs.com/package/typescript

pdurbin commented:
@kcondon I had the same problem: https://github.com/IQSS/dataverse-client-javascript/pull/55#discussion_r1163265878

In the end the solution was `npm install`.

kcondon commented:
> To test this functionality we have to import the npm package from a test application. Usually we would install the package from the npm registry, but since it is a development unpublished version, we have to reference a pre-built local package from the test application.
> 
> To build the package locally:
> 
> * 1. Install dependencies omitting dev: `npm install --omit=dev`.
> * 2. Build the package code: `npm run build`.
> * 3. Create the .tgz local package: `npm pack`.

Hi, @tainguyenbui Installing typescript allowed me to build npm, thanks! I also needed to install jest to run the tests.

kcondon commented:
> @kcondon I had the same problem: [#55 (comment)](https://github.com/IQSS/dataverse-client-javascript/pull/55#discussion_r1163265878)
> 
> In the end the solution was `npm install`.

What about the npm build instructions in the testing section then? Normally I would just run npm install to start but I was assuming that would fetch some preexisting version and the special build instructions would do something different or is that not the case?

pdurbin reviewed (approved):
I thought I clicked approve already. Whoops! Doing so now.
